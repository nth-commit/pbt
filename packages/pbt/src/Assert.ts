import { Property } from 'pbt-properties';
import { run, RunConfig, RunResult } from './Run';

export type AssertionJournal = string[];

export const assert = <Values extends any[]>(p: Property<Values>, config?: Partial<RunConfig>): AssertionJournal => {
  const result = run(p, config);
  switch (result.kind) {
    case 'success':
      return [];
    case 'failure':
      return [...renderTitle(result), ...renderReproduction(result), ...renderCounterexample(result)];
    /* istanbul ignore next */
    default:
      throw new Error(`Unhandled result: ${JSON.stringify(result)}`);
  }
};

const renderTitle = (result: RunResult.Failure): string[] => [
  `Property failed after ${result.iterationsCompleted} test(s)`,
];

const renderReproduction = (result: RunResult.Failure): string[] => {
  let str = `Reproduction: { "seed": ${result.seed}, "size": ${result.size}`;

  if (result.counterexample.shrinkPath) {
    str += `, "shrinkPath": "${result.counterexample.shrinkPath}"`;
  }

  str += ` }`;

  return [str];
};

const renderCounterexample = (result: RunResult.Failure): string[] => [
  `Counterexample: ${JSON.stringify(result.counterexample.values)}`,
];
