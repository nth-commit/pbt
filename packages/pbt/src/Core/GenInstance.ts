/* istanbul ignore file */
import { pipe, toArray, from } from 'ix/iterable';
import { flatMap, map } from 'ix/iterable/operators';

export type GenInstanceData<T> = {
  readonly value: T;
  shrink(): Iterable<GenInstanceData<T>>;
};

export type GenInstance<T> = {
  kind: 'instance';
} & GenInstanceData<T>;

export type GenInstances = GenInstance<any>[];

type GenInstanceDatas = GenInstanceData<any>[];

type GenInstanceValue<T> = T extends GenInstanceData<infer U> ? U : never;

type GenInstanceValues<TGenInstanceDatas extends GenInstanceDatas> = {
  [P in keyof TGenInstanceDatas]: GenInstanceValue<TGenInstanceDatas[P]>;
};

export type GenInstanceMapper<TGenInstances extends GenInstances> = (
  ...args: GenInstanceValues<TGenInstances>
) => boolean;

export type EvaluatedInstance<T> = T | [T, Array<EvaluatedInstance<T>>];

export namespace GenInstance {
  type EvaluatedInstanceComplex<T> = [T, Array<EvaluatedInstanceComplex<T>>];

  const evaluateInstanceComplex = <T>(instance: GenInstanceData<T>): EvaluatedInstanceComplex<T> => {
    const evaluatedShrinks = toArray(
      pipe(
        instance.shrink(),
        map((i) => evaluateInstanceComplex(i)),
      ),
    );
    return [instance.value, evaluatedShrinks];
  };

  const simplifyEvaluation = <T>([x, xs]: EvaluatedInstanceComplex<T>): EvaluatedInstance<T> => {
    if (xs.length === 0) {
      return x;
    }
    return [x, xs.map(simplifyEvaluation)];
  };

  export const evaluate = <T>(instance: GenInstanceData<T>): EvaluatedInstance<T> =>
    simplifyEvaluation(evaluateInstanceComplex(instance));

  function unfoldData<Seed, T>(f: (x: Seed) => T, g: (x: Seed) => Iterable<Seed>, x: Seed): GenInstanceData<T> {
    return {
      value: f(x),
      shrink: () => unfoldDatas(f, g, x),
    };
  }

  function unfoldDatas<Seed, T>(
    f: (x: Seed) => T,
    g: (x: Seed) => Iterable<Seed>,
    x: Seed,
  ): Iterable<GenInstanceData<T>> {
    return pipe(
      g(x),
      map((y) => unfoldData(f, g, y)),
    );
  }

  export const unfold = <Seed, T>(f: (x: Seed) => T, g: (x: Seed) => Iterable<Seed>, x: Seed): GenInstance<T> => ({
    kind: 'instance',
    ...unfoldData(f, g, x),
  });

  const joinDatas = <TGenInstanceDatas extends GenInstanceDatas>(
    ...instanceDatas: TGenInstanceDatas
  ): GenInstanceData<GenInstanceValues<TGenInstanceDatas>> => ({
    value: instanceDatas.map((d) => d.value) as GenInstanceValues<TGenInstanceDatas>,
    shrink: () =>
      pipe(
        from(instanceDatas),
        map((instanceData, i) => {
          const leftInstanceDatas = instanceDatas.slice(0, i);
          const rightInstanceDatas = instanceDatas.slice(i + 1);
          return pipe(
            instanceData.shrink(),
            map((instanceDataShrink) => joinDatas(...leftInstanceDatas, instanceDataShrink, ...rightInstanceDatas)),
          );
        }),
        flatMap((x) => x),
      ) as Iterable<GenInstanceData<GenInstanceValues<TGenInstanceDatas>>>,
  });

  export const join = <TGenInstances extends GenInstances>(
    ...instances: TGenInstances
  ): GenInstance<GenInstanceValues<TGenInstances>> => ({
    kind: 'instance',
    ...joinDatas(...instances),
  });

  const formatInternal = (instance: GenInstanceData<string>, nestCount: number): string => {
    const valueFormatted = '-'.repeat(nestCount * 3) + instance.value;

    const shrinksFormatted = toArray(
      pipe(
        instance.shrink(),
        map((i) => formatInternal(i, nestCount + 1)),
      ),
    );

    return [valueFormatted, ...shrinksFormatted].join('\n');
  };

  export const format = (instance: GenInstanceData<string>): string => formatInternal(instance, 0);
}
