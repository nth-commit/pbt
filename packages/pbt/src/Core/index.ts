export * from './Seed';
export * from './Size';
export * from './Tree';
export * from './RandomStream';
