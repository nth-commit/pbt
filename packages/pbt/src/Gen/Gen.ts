import { GenTree } from '../GenTree';
import { Gen as G, GenFactory, ArrayGen, IntegerGen, ElementGen, StateMachineGen, PrimitiveGen } from './Abstractions';
import { GenIteration } from './GenIteration';
import { array, integer, element, primitive, stateMachine } from './Gens';
import { RawGenImpl } from './Gens/RawGenImpl';
import { Shrink } from './Shrink';

const genFactory: GenFactory = {
  primitive: (generate, shrink, measure) => primitive(generate, shrink, measure, genFactory),
  constant: (value) => genFactory.primitive(() => value, Shrink.none(), GenTree.CalculateComplexity.none()),
  error: (message: string) =>
    RawGenImpl.fromRunFunction((rng, size) => [GenIteration.error(message, rng, rng, size, size)], genFactory),
  integer: () => integer(genFactory),
  array: (elementGen) => array(elementGen, genFactory),
  element: (collection) => element(collection, genFactory),
  stateMachine: (initialState, generateTransition, applyTransition) =>
    stateMachine(genFactory, initialState, generateTransition, applyTransition),
};

export type Gen<T> = G<T>;

export type Gens<Ts extends any[]> = { [P in keyof Ts]: Gen<Ts[P]> };

export namespace Gen {
  export type StatefulGenFunction<T> = PrimitiveGen.StatefulGenFunction<T>;
  export type NextIntFunction = PrimitiveGen.NextIntFunction;

  /**
   * Creates a generator from a series of functions.
   *
   * @param generate
   * @param shrink
   * @param measure
   */
  export const create = <T>(
    generate: StatefulGenFunction<T>,
    shrink: Shrink<T>,
    measure: GenTree.CalculateComplexity<T> = GenTree.CalculateComplexity.none(),
  ): Gen<T> => genFactory.primitive(generate, shrink, measure);

  /**
   * Creates a generator that always returns the given value, and does not shrink.
   *
   * @param value
   */
  export const constant = <T>(value: T): Gen<T> => genFactory.constant(value);

  /**
   * Creates a generator that produces a single error signal, then terminates. This is the recommended way to exit from
   * functions that produce generators if the arguments were invalid, as the error signal will be appropriately
   * handled by the runners.
   *
   * @param message
   */
  /* istanbul ignore next */
  export const error = <T>(message: string): Gen<T> => genFactory.error(message);

  /**
   * Creates a generator for integers.
   */
  export const integer = (): IntegerGen => genFactory.integer();

  /**
   * Creates a generator for arrays, where the elements of the array are values from the given generator.
   *
   * @param elementGen
   */
  export const array = <T>(elementGen: Gen<T>): ArrayGen<T> => genFactory.array(elementGen);

  /**
   * Creates a generator for tuples, where the elements of the tuple are values from the given generators, and are
   * placed in the resultant tuple position-wise.
   *
   * @param gens
   */
  export const zip = <Ts extends any[]>(...gens: Gens<Ts>): Gen<Ts> => {
    switch (gens.length) {
      case 0: {
        return constant(([] as unknown) as Ts);
      }
      case 1: {
        const [gen] = gens;
        return gen.map((x) => [x] as Ts);
      }
      default: {
        const [gen, ...nextGens] = gens;
        return gen.flatMap((x) => zip(...nextGens).map((xs) => [x, ...xs] as Ts));
      }
    }
  };

  export type VariadicMapper<Ts extends any[], U> = (...xs: { [P in keyof Ts]: Ts[P] }) => U;

  /**
   * Creates a generator, where the values of the given generators are provided as arguments for the mapper function,
   * and the results of the mapper function become the values for the resultant generator.
   *
   * @example Gen2.map(Gen2.constant('Hello'), Gen2.constant('World'), (greeting, subject) => `${greeting}, ${subject}`);
   *
   * @param args
   */
  export const map = <Ts extends any[], U>(...args: [...Gens<Ts>, VariadicMapper<Ts, U>]): Gen<U> => {
    const [f, ...gens] = args.reverse();
    const fUnsafe = f as VariadicMapper<any[], U>;
    const gensUnsafe = gens as Gens<any[]>;
    return zip(...gensUnsafe).map(([...xs]) => fUnsafe(...xs));
  };

  /**
   * Creates a generator, where the values of the resultant generator are a random element from the input collection.
   *
   * @param collection
   */
  export const element = <T>(collection: ElementGen.Collection<T>): ElementGen<T> => genFactory.element(collection);

  export type GenerateTransitionFunction<State, Transition> = StateMachineGen.GenerateTransitionFunction<
    State,
    Transition
  >;

  export type ApplyTransitionFunction<State, Transition> = StateMachineGen.ApplyTransitionFunction<State, Transition>;

  /**
   * Creates a generator, where the values are states of a finite state machine. The state machine is defined by two
   * functions, the first produces
   * @param initialState
   * @param generateTransition
   * @param applyTransition
   */
  export const stateMachine = <State, Transition>(
    initialState: State,
    generateTransition: GenerateTransitionFunction<State, Transition>,
    applyTransition: ApplyTransitionFunction<State, Transition>,
  ): StateMachineGen<State> => genFactory.stateMachine(initialState, generateTransition, applyTransition);
}
