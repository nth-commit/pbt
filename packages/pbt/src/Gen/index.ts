export * from './Imports';
export * from './Shrink';
export * from './Gen';
export * from './Operators';
export * from './Number';
export * from './Collections';
