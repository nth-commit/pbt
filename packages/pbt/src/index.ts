export { RandomStream, Seed, Size } from './Core';
export { GenTree, GenTreeNode, Complexity, CalculateComplexity } from './GenTree';
export { Gen, Shrink, Shrinker, GenIteration } from './Gen';
export { property, Property, PropertyIteration, PropertyFunction, ShrinkIteration, Counterexample } from './Property';
export {
  check,
  CheckConfig,
  CheckResult,
  sample,
  sampleTrees,
  SampleConfig,
  Sample,
  assert,
  AssertConfig,
  PbtAssertionError,
} from './Runners';
