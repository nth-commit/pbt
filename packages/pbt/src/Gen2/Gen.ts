/* istanbul ignore file */

import { Seed, Size } from '../Core';
import { CalculateComplexity } from '../GenTree';
import { GenFunction } from './GenFunction';
import { Shrinker } from './Shrink';
import { Gen as _Gen, ArrayGen, IntegerGen, GenFactory } from './Abstractions';
import { BaseGen } from './BaseGen';
import { array } from './ArrayGen';
import { integer } from './IntegerGen';

export const genFactory: GenFactory = {
  array: (elementGen) => array(elementGen, genFactory),
  integer: () => integer(genFactory),
};

export type Gen<T> = _Gen<T>;

export type AnyValues = any[];

export type Gens<Values extends AnyValues> = { [P in keyof Values]: Gen<Values[P]> };

export namespace Gen {
  export function create<T>(
    generate: (seed: Seed, size: Size) => T,
    shrink: Shrinker<T>,
    calculateComplexity: CalculateComplexity<T>,
  ): Gen<T> {
    return new BaseGen(GenFunction.create(generate, shrink, calculateComplexity), genFactory);
  }

  export function constant<T>(x: T): Gen<T> {
    return new BaseGen(GenFunction.constant(x), genFactory);
  }

  export function array<T>(elementGen: Gen<T>): ArrayGen<T> {
    return genFactory.array(elementGen);
  }

  export function integer(): IntegerGen {
    return genFactory.integer();
  }

  export function zip<Ts extends any[]>(...gens: Gens<Ts>): Gen<Ts> {
    switch (gens.length) {
      case 0:
        return constant(([] as unknown) as Ts);
      default: {
        const [gen, ...nextGens] = gens;
        return gen.flatMap((x) => zip(...nextGens).map((xs) => ([x, ...xs] as unknown) as Ts));
      }
    }
  }

  export type MapperFunction<Ts extends any[], U> = (...xs: { [TLabel in keyof Ts]: Ts[TLabel] }) => U;
  export type MapArgs<Ts extends any[], U> = [...Gens<Ts>, MapperFunction<Ts, U>];

  export function map<Ts extends any[], U>(...args: MapArgs<Ts, U>): Gen<U> {
    const [f, ...gens] = args.reverse();
    const fUnsafe = f as MapperFunction<any[], U>;
    const gensUnsafe = gens as Gens<any[]>;
    return zip(...gensUnsafe).map(([...xs]) => fUnsafe(...xs));
  }
}
