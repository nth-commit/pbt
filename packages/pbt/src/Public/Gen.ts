import { Seed, Size } from '../Core';
import * as Internal from '../Gen';
import { RandomStream } from './RandomStream';

const mapExternalKToInternalK = <T, U>(k: (x: T) => Gen<U>): ((x: T) => Internal.Gen<U>) => {
  return (x) => {
    const external = k(x);
    return external.genFunction;
  };
};

export type Shrinker<T> = Internal.Shrinker<T>;

export class Gen<T> implements RandomStream<Internal.GenIteration<T>> {
  static create = <T>(f: (seed: Seed, size: Size) => T, shrinker: Internal.Shrinker<T>): Gen<T> =>
    new Gen(Internal.create(f, shrinker));

  static integer = {
    unscaled: (min: number, max: number): Gen<number> => new Gen(Internal.integer.unscaled(min, max)),
    scaleLinearly: (min: number, max: number): Gen<number> => new Gen(Internal.integer.scaleLinearly(min, max)),
  };

  static naturalNumber = {
    unscaled: (max?: number): Gen<number> => new Gen(Internal.naturalNumber.unscaled(max)),
    scaleLinearly: (max?: number): Gen<number> => new Gen(Internal.naturalNumber.scaleLinearly(max)),
  };

  static array = {
    unscaled: <T>(min: number, max: number, gen: Gen<T>): Gen<T[]> =>
      new Gen<T[]>(Internal.array.unscaled(min, max, gen.genFunction)),
    scaleLinearly: <T>(min: number, max: number, gen: Gen<T>): Gen<T[]> =>
      new Gen<T[]>(Internal.array.scaleLinearly(min, max, gen.genFunction)),
  };

  static element = <T>(collection: Array<T> | Record<any, T> | Set<T> | Map<any, T>): Gen<T> =>
    new Gen(Internal.element(collection));

  private constructor(public readonly genFunction: Internal.Gen<T>) {}

  map<U>(f: (x: T) => U): Gen<U> {
    return new Gen<U>(Internal.map(this.genFunction, f));
  }

  flatMap<U>(k: (x: T) => Gen<U>): Gen<U> {
    return new Gen<U>(Internal.flatMap(this.genFunction, mapExternalKToInternalK(k)));
  }

  filter(predicate: (x: T) => boolean): Gen<T> {
    return new Gen<T>(Internal.filter(this.genFunction, predicate));
  }

  reduce<U>(length: number, f: (acc: U, x: T, i: number) => U, init: U): Gen<U> {
    return new Gen<U>(Internal.reduce(this.genFunction, length, f, init));
  }

  noShrink(): Gen<T> {
    return new Gen<T>(Internal.noShrink(this.genFunction));
  }

  postShrink(shrinker: Shrinker<T>): Gen<T> {
    return new Gen<T>(Internal.postShrink(this.genFunction, shrinker));
  }

  run(seed: Seed, size: Size): Iterable<Internal.GenIteration<T>> {
    return this.genFunction(seed, size);
  }
}
