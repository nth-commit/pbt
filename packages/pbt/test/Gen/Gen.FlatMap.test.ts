import fc from 'fast-check';
import * as devCore from '../../src/Core';
import * as dev from '../../src/Gen';
import * as domainGen from './Helpers/domainGen';
import { iterate, iterateTrees } from './Helpers/genRunner';

test('It discards when the bound gen discards', () => {
  fc.assert(
    fc.property(domainGen.runParams(), domainGen.firstOrderGen(), (runParams, baseGen) => {
      const gen = dev.operators.flatMap(baseGen, () => dev.operators.filter(baseGen, () => false));

      const genIterations = iterate(gen, runParams);

      genIterations.forEach((genIteration) => {
        expect(genIteration.kind).toEqual('discard');
      });
    }),
  );
});

test('It exhausts when the bound gen exhausts', () => {
  fc.assert(
    fc.property(domainGen.runParams(), domainGen.firstOrderGen(), (runParams, baseGen) => {
      const gen = dev.operators.flatMap(baseGen, () => dev.exhausted());

      const genIterations = iterate(gen, runParams);

      expect(genIterations).toEqual([{ kind: 'exhausted' }]);
    }),
  );
});

test('Snapshot', () => {
  // It's hard to describe the intricacies of how flatMap combines streams. It's some kind of join or summit. Check the
  // snapshots, yo!

  const seed = dev.Seed.create(0);
  const iterationsBySize = new Map<number, number>([
    [0, 1],
    [20, 2],
    [50, 5],
  ]);

  const unflatMappedGen = dev.integer.scaleLinearly(0, 5);
  const flatMappedGen = dev.operators.flatMap(unflatMappedGen, (x) =>
    dev.operators.map(unflatMappedGen, (y) => `[${x},${y}]`),
  );

  for (const [size, iterations] of iterationsBySize.entries()) {
    iterateTrees(flatMappedGen, { seed, size, iterations })
      .map(devCore.Tree.format)
      .forEach((result, i) => expect(result).toMatchSnapshot(`size=${size} iteration=${i + 1}`));
  }
});
