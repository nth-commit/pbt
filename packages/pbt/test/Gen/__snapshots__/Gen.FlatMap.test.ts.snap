// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: size=0 iteration=1 1`] = `"[0,0]"`;

exports[`Snapshot: size=20 iteration=1 1`] = `
"[0,1]
---[0,0]"
`;

exports[`Snapshot: size=20 iteration=2 1`] = `
"[1,0]
---[0,0]"
`;

exports[`Snapshot: size=50 iteration=1 1`] = `
"[0,1]
---[0,0]"
`;

exports[`Snapshot: size=50 iteration=2 1`] = `
"[1,2]
---[0,2]
------[0,0]
------[0,1]
---------[0,0]
---[1,0]
---[1,1]
------[1,0]"
`;

exports[`Snapshot: size=50 iteration=3 1`] = `
"[0,2]
---[0,0]
---[0,1]
------[0,0]"
`;

exports[`Snapshot: size=50 iteration=4 1`] = `
"[3,2]
---[0,2]
------[0,0]
------[0,1]
---------[0,0]
---[1,2]
------[0,2]
---------[0,0]
---------[0,1]
------------[0,0]
------[1,0]
------[1,1]
---------[1,0]
---[2,2]
------[0,2]
---------[0,0]
---------[0,1]
------------[0,0]
------[1,2]
---------[0,2]
------------[0,0]
------------[0,1]
---------------[0,0]
---------[1,0]
---------[1,1]
------------[1,0]
------[2,0]
------[2,1]
---------[2,0]
---[3,0]
---[3,1]
------[3,0]"
`;

exports[`Snapshot: size=50 iteration=5 1`] = `
"[0,3]
---[0,0]
---[0,1]
------[0,0]
---[0,2]
------[0,0]
------[0,1]
---------[0,0]"
`;
